# Homework 1: Introduction to Machine Learning 

### Q1. Pandas version

In [1]:
import pandas as pd

pd.__version__

'3.0.5'

The version of pandas installed in this notebook is: `3.0.5`

### Getting the data

In [2]:
!wget https://raw.githubusercontent.com/DataTalksClub/machine-learning-zoomcamp/main/cohorts/2026/data/car_fuel_efficiency_2026.csv



7[Files: 0  Bytes: 0  [0 B/s] Re]87[https://raw.githubusercontent.]87Saving 'car_fuel_efficiency_2026.csv'
87car_fuel_efficiency_  28% [=======>                      ]   33.30K    --.-KB/s87[Files: 0  Bytes: 0  [0 B/s] Re]87car_fuel_efficiency_ 100% [=============================>]  116.79K  684.33KB/s87HTTP response 200  [https://raw.githubusercontent.com/DataTalksClub/machine-learning-zoomcamp/main/cohorts/2026/data/car_fuel_efficiency_2026.csv]
87car_fuel_efficiency_ 100% [=============================>]  116.79K  684.33KB/s87[Files: 1  Bytes: 116.79K [104.]8

In [3]:
df = pd.read_csv('car_fuel_efficiency_2026.csv')

In [4]:
df.head()

,model_year,origin,fuel_type,drivetrain,num_doors,engine_displacement,num_cylinders,horsepower,vehicle_weight,acceleration,fuel_efficiency_mpg
0,2006,Europe,Gasoline,Front-wheel drive,4,2180,6,243.0,3870,NaN,31.9
1,2008,Europe,Diesel,Front-wheel drive,4,2390,6,272.0,4210,NaN,31.3
2,1996,Asia,Gasoline,Front-wheel drive,5,2320,6,267.0,4240,17.3,27.5
3,1989,Europe,Gasoline,Front-wheel drive,4,2130,6,258.0,4490,18.7,28.5
4,1994,USA,Diesel,Front-wheel drive,3,2580,7,304.0,4510,17.5,31.0


### Q2. Records count

In [5]:
df.shape

(10000, 11)

There are 10000 records in this dataset.

### Q3. Fuel types

In [6]:
df['fuel_type'].unique()

<StringArray>
['Gasoline', 'Diesel', 'Hybrid']
Length: 3, dtype: str

3 fuel type are presented in this dataset.

### Q4. Missing values

In [10]:
df.isna().sum()

model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors                0
engine_displacement      0
num_cylinders            0
horsepower             877
vehicle_weight           0
acceleration           264
fuel_efficiency_mpg      0
dtype: int64

2 columns have missing values

### Q5. Max fuel efficiency

In [53]:
df[df['origin'] == 'Asia']['fuel_efficiency_mpg'].max()

np.float64(41.2)

The maximum fuel efficiency of cars from Asia is 41.2

### Q6. Median value of horsepower

1. Find the median value of the `horsepower` column in the dataset.

In [13]:
pd.DataFrame(df['horsepower'].describe()).T

,count,mean,std,min,25%,50%,75%,max
horsepower,9123.0,254.446016,22.844613,171.0,239.0,254.0,270.0,334.0


2. Next, calculate the most frequent value of the same `horsepower` column.

In [14]:
df['horsepower'].value_counts()

horsepower
252.0    174
246.0    168
256.0    168
259.0    165
253.0    165
        ... 
322.0      1
330.0      1
175.0      1
178.0      1
320.0      1
Name: count, Length: 153, dtype: int64

3. Use the fillna method to fill the missing values in the `horsepower` column with the most frequent value from the previous step.

In [20]:
df['horsepower'] = df['horsepower'].fillna(252.0)

In [21]:
df['horsepower'].isna().sum()

np.int64(0)

4. Now, calculate the median value of `horsepower` once again.

In [23]:
pd.DataFrame(df['horsepower'].describe()).T

,count,mean,std,min,25%,50%,75%,max
horsepower,10000.0,254.2315,21.830757,171.0,241.0,252.0,268.0,334.0


Yes, it decreased

### Q7. Sum of weights

1. Select all the cars from Asia

In [24]:
df_asia = df[df['origin'] == 'Asia']

In [25]:
df_asia.head()

,model_year,origin,fuel_type,drivetrain,num_doors,engine_displacement,num_cylinders,horsepower,vehicle_weight,acceleration,fuel_efficiency_mpg
2,1996,Asia,Gasoline,Front-wheel drive,5,2320,6,267.0,4240,17.3,27.5
6,1992,Asia,Gasoline,Front-wheel drive,2,2270,6,250.0,4310,21.7,26.8
8,1997,Asia,Gasoline,Rear-wheel drive,3,2330,6,251.0,4230,19.5,30.9
10,1978,Asia,Gasoline,All-wheel drive,3,2410,6,252.0,4420,19.5,26.8
15,2007,Asia,Diesel,Front-wheel drive,4,2170,5,226.0,3950,18.0,35.4


2. Select only columns `vehicle_weight` and `model_year`

In [28]:
df_filtered_columns = df_asia[['vehicle_weight', 'model_year']]

In [29]:
df_filtered_columns.head()

,vehicle_weight,model_year
2,4240,1996
6,4310,1992
8,4230,1997
10,4420,1978
15,3950,2007


3. Select the first 7 values

In [31]:
df_seven = df_filtered_columns.head(7)

4. Get the underlying NumPy array. Let's call it `X`.

In [33]:
X = df_seven.to_numpy()
X

array([[4240, 1996],
       [4310, 1992],
       [4230, 1997],
       [4420, 1978],
       [3950, 2007],
       [4750, 2021],
       [4450, 2019]])

5. Compute matrix-matrix multiplication between the transpose of `X` and X. To get the transpose, use `X.T`. Let's call the result `XTX`

In [34]:
XTX = X.T @ X

In [35]:
XTX

array([[131950500,  60750580],
       [ 60750580,  28041424]])

5. Invert `XTX`

In [36]:
import numpy as np

In [37]:
XTX_inverted = np.linalg.inv(XTX)

6. Create an array `y` with values `[1100, 1300, 800, 900, 1000, 1100, 1200]`.

In [38]:
y = np.array([[1100, 1300, 800, 900, 1000, 1100, 1200]])

8. Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.

In [42]:
w = (XTX_inverted @ X.T) @ y.T
w

array([[0.13644777],
       [0.2327492 ]])

9. What's the sum of all the elements of the result?

In [43]:
w.sum()

np.float64(0.36919696904925486)

The sum of all elements of the result is: `0.369`

---